### PR TITLE
Verify that timestamp is before the signing certificate expiry

### LINKED ISSUE #35: Verify that timestamp is before the signing certificate expiry
We should check the signing timestamp that Rekor gives us and ensure that it is before the expiry of the signing certificate.

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
Closes #35

### CODE REVIEW

File: sigstore/_verify.py [hunk 1]
```diff
@@ -148,6 +151,18 @@ def verify(
...
+        if (
+            integrated_time < cert.not_valid_before
+            or integrated_time > cert.not_valid_after
+        ):
+            # No need to log anything here.
+            #
+            # If an artifact has been signed multiple times, this will happen so it's not really an
+            # error case.
+            continue
+
```
Comment (woodruffw) [inline sigstore/_verify.py:165]: I might be misunderstanding the behavior here -- does this actually verify the timestamp? It seems as though it just `continue`s if the timestamp is invalid.

Comment (di) [reply sigstore/_verify.py:165]: IIUC, this is looking at every matching entry, and checking if any of the entries match the window. If none of them match, this loop will never get to the `valid_sig_exists = True` line and the validation will fail.

File: sigstore/_verify.py [hunk 2]
```diff
@@ -148,6 +151,18 @@ def verify(
             output(f"Failed to validate Rekor entry's SET: {inval_set}")
             continue
 
+        # 6) Verify that the signing certificate was valid at the time of signing
+        integrated_time = datetime.datetime.utcfromtimestamp(entry.integrated_time)
+        if (
+            integrated_time < cert.not_valid_before
+            or integrated_time > cert.not_valid_after
```
Comment (di) [inline sigstore/_verify.py:158]: Just want to confirm that neither of these should be inclusive? Specifically what should happen when `integrated_time == cert.not_valid_before`?

Comment (woodruffw) [reply sigstore/_verify.py:158]: RFC 5280 isn't completely explicit about this, but I believe `notBefore` is inclusive and `notAfter` is exclusive. 

Quote https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.5:

> The field is represented as a SEQUENCE of two dates:
   the date on which the certificate validity period begins (notBefore)
   and the date on which the certificate validity period ends
   (notAfter).  Both notBefore and notAfter may be encoded as UTCTime or
   GeneralizedTime.

My reading of that is: "the certificate validity begins with `notBefore` (meaning inclusive) and ends with `notAfter` (meaning exclusive)".

Comment (woodruffw) [reply sigstore/_verify.py:158]: Whoops, my language there is confusing: `time == notBefore` is valid and `time == notAfter` is invalid, based on my interpretation.

Comment (woodruffw) [reply sigstore/_verify.py]: [d574759](https://github.com/trailofbits/sigstore-python/pull/42/commits/d574759123981512c7ab2f72172943b74c09931e)

**Review by woodruffw — APPROVED** [review]

File: sigstore/_verify.py (hunk 2 above)
Comment (tetsuo-cpp) [reply sigstore/_verify.py]: Good eye. Thanks.

Comment (di) [reply sigstore/_verify.py]: Looks like d574759 does the opposite of what I'd expect?

Comment (di) [reply sigstore/_verify.py]: Oh, this is the failure case. 👍

**Comment by woodruffw** [comment]
Yep, that’s also what threw me off when describing it 🙂

**Pull request merged**